### PR TITLE
pkg: add request IDs to response/logs

### LINKED ISSUE #249: Add Request ID to Requests
### What is the problem/pain point?

- It is currently difficult to troubleshoot customer requests given that we don't capture any easily identifiable information that we can relate back to a particular request

### Who is impacted?

- Customers onboarding and troubleshooting their configuration

### What is the impact?

- Troubleshooting takes quite a while, which takes time away from engineering tasks

### Why now?

- As we onboard more customers we need to be able to troubleshoot individual customer issues faster

### User Story
- As a Storj Solutions Architect, I want to be able to identify individual authservice and linksharing requests so I can quickly identify and troubleshoot specific customer problems.

### Acceptance Criteria

- Request ID is present on linksharing and authservice request logs

### Success Metrics

- Time to resolution is faster
- Less need for engineering help

### PR DESCRIPTION
**Pull request by aishwaryasharmaccoew**
Adds X-Storj-Request-Id field to response headers and logs it; for
gateway-mt, linksharing, and authservice. This makes it easier to debug
these services (especially) for our customers.

## PR Checklist

- [x] The PR title is formatted as follows: `pkg/server: add TestXYZ for XYZ`
  - [x] The package name goes before the colon
  - [x] The part after the colon uses the verb tense + phrase that completes the blank in, "This change modifies Gateway-MT to ___________"
  - [x] Lowercase verb after the colon
  - [x] No trailing period
  - [x] Keep the title as short as possible. Ideally under 72 characters or shorter
  - [x] No Markdown

## Related Issue

Closes #249

## Code Review Checklist (to be filled out by reviewer)

- [x] Are there any authservice's database migrations? Are they forwards _and_ backwards compatible?
- [x] Does the PR describe what changes are being made?
- [x] Does the PR describe why the changes are being made?
- [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/main/code/Style.md)?
- [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/main/code/Testing.md)?
- [x] Is the PR appropriately sized? (If it could be broken into smaller PRs, it should be)
- [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
- [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
- [x] Does any documentation needs updating?
- [x] Do the database access patterns make sense?
- [x] Copy the Commit Message Body section contents into the submit prompt upon PR completion

### CODE REVIEW

File: pkg/server/middleware/handler_utils.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+// UTCNow - returns current UTC time.
+func utcNow() time.Time {
+	return time.Now().UTC()
+}
+
+func AddRequestIds(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Reusing the field x-amz-request-id to set unique request Ids for each request.
+		w.Header().Set(xhttp.AmzRequestID, mustGetRequestID(utcNow()))
```
Comment (amwolff) [inline pkg/server/middleware/handler_utils.go:23]: It's probably best if we use our custom header like `X-Storj-Request-Id`.

Comment (amwolff) [inline pkg/server/middleware/handler_utils.go:23]: How about if, for requests, we used IDs that monkit generates for tracing?

File: pkg/auth/peer.go [hunk 2]
```diff
@@ -203,12 +206,14 @@ func New(ctx context.Context, log *zap.Logger, config Config, configDir string)
 // LogRequests logs requests.
 func LogRequests(log *zap.Logger, h http.Handler) http.Handler {
 	return whroute.HandlerFunc(h, func(w http.ResponseWriter, r *http.Request) {
+
```
Comment (halkyon) [inline pkg/auth/peer.go:209]: nit: try to reduce these whitespace changes for code that isn't being changed, or make them a separate commit

File: pkg/server/middleware/handler_utils.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+// AddRequestIds uses XStorjRequestId field to set unique request Ids
+// in the response headers for each request of auth and linksharing service.
+// It also mantains calling service request Id in the XStorjParentRequestId field in auth service response header.
+func AddRequestIds(service string, h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Storing calling service requestId as ParentRequestID for auth service
+		if service == "auth" {
+			if w.Header().Get(XStorjRequestID) != "" {
+				w.Header().Set(XStorjParentRequestID, w.Header().Get(XStorjRequestID))
```
Comment (halkyon) [inline pkg/server/middleware/handler_utils.go:30]: if we already have a `X-Storj-Request-Id` in the request headers, shouldn't we pass that straight through instead of putting the existing one in parent? As far as I understand, this middleware should only be generating an ID and setting `X-Storj-Request-Id` if the header is missing, and then setting the one from the request, or the one it just generated to the response header `X-Storj-Request-Id`. I'm not sure what the idea is with the parent header, but maybe I didn't pick up on how this should work.

I would also remove the service check and have this work the same way regardless of who called it.

File: pkg/server/middleware/handler_utils.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
+// Copyright (C) 2021 Storj Labs, Inc.
```
Comment (halkyon) [inline pkg/server/middleware/handler_utils.go:1]: nit: copyright should be 2022.

Maybe we can also call these files `requestid`? `handler_utils` is probably a bit too generic.

File: pkg/server/middleware/handler_utils_test.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
+// Copyright (C) 2021 Storj Labs, Inc.
```
Comment (halkyon) [inline pkg/server/middleware/handler_utils_test.go:1]: nit: 2022 copyright here as well

File: pkg/server/middleware/handler_utils.go (hunk 3 above)
Comment (amwolff) [reply pkg/server/middleware/handler_utils.go:30]: Oh, that's a good idea. I suggested parent request IDs to Aishwarya, so sorry about the confusion. I think Sean's proposal is definitely better.

File: pkg/server/middleware/handler_utils.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+			if w.Header().Get(XStorjRequestID) != "" {
+				w.Header().Set(XStorjParentRequestID, w.Header().Get(XStorjRequestID))
+			} else if w.Header().Get(xhttp.AmzRequestID) != "" {
+				w.Header().Set(XStorjParentRequestID, w.Header().Get(xhttp.AmzRequestID))
+			} else {
+				w.Header().Set(XStorjParentRequestID, "")
+			}
+		}
+
+		trace := monkit.NewTrace(monkit.NewId())
```
Comment (amwolff) [inline pkg/server/middleware/handler_utils.go:38]: I think we might need to fuse this handler with where traces are now created. What do you think?

File: pkg/httpserver/log.go [hunk 7]
```diff
@@ -60,6 +60,7 @@ func logResponses(log *zap.Logger, h http.Handler) http.Handler {
 				// we are deliberately not logging the request URI as it has
 				// sensitive information in it.
 				zap.Int("code", code),
+				zap.String("request-id", rw.Header().Get("X-Storj-Request-Id")),
```
Comment (halkyon) [inline pkg/httpserver/log.go:63]: this might be better as `GetReqID(ctx)` then you don't have to hardcode the header name

File: pkg/pkgmiddleware/requestId.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+	// RequestIDKey is the key that holds the unique request ID in a request context.
+	RequestIDKey ctxKeyRequestID = "request-id"
+)
+
+// AddRequestID uses XStorjRequestId field to set unique request Ids
+// in the response headers for each request of auth and linksharing service, if it dosen't alreasy exist.
+func AddRequestID(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		if w.Header().Get(XStorjRequestID) == "" {
```
Comment (halkyon) [inline pkg/pkgmiddleware/requestId.go:30]: i think this should be `r.Header().Get(...` because we want to take the ID from an incoming request if it was given. I was thinking something like:

```
ctx := r.Context()
if r.Header().Get(XStorjRequestID) != "" {
  ctx = context.WithValue(ctx, RequestIDKey, r.Header().Get(XStorjRequestID))
  h.ServeHTTP(w, r.WithContext(ctx))
  return
}

trace := monkit.NewTrace(monkit.NewId())
requestID := fmt.Sprintf("%x", trace.Id())
w.Header().Set(XStorjRequestID, requestID)
ctx = context.WithValue(ctx, RequestIDKey, requestID)
h.ServeHTTP(w, r.WithContext(ctx))
```

File: pkg/server/middleware/handler_utils.go (hunk 6 above)
Comment (amwolff) [reply pkg/server/middleware/handler_utils.go]: We talked about this with Sean, and it's probably too much, for now, so I'll resolve this. Let's have duplicate IDs for both logs and tracing for now, and we can clean this up later.

File: pkg/pkgmiddleware/requestId.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/spacemonkeygo/monkit/v3"
+)
+
+// Key to use when setting the request ID.
+type ctxKeyRequestID string
```
Comment (amwolff) [inline pkg/pkgmiddleware/requestId.go:15]: nit: a more robust context key would be something like this:

```suggestion
type RequestIDKey struct{}
```

then you can just:

```
... context.WithValue(ctx, RequestIDKey{}, ...) ...
```

and then:

```
... ctx.Value(RequestIDKey{}).(...) ...
```

File: pkg/authclient/authclient.go [hunk 10]
```diff
@@ -59,6 +60,8 @@ func (a *AuthClient) Resolve(ctx context.Context, accessKeyID string, clientIP s
 	}
 	for {
 		resp, err := client.Do(req)
+		middleware.AddReqIDHeader(ctx, resp)
```
Comment (halkyon) [inline pkg/authclient/authclient.go:63]: I think you want to move this up to req creation and add the header there, not to `resp`.

File: pkg/middleware/requestId.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+const (
+	// XStorjRequestID is the header key for the request ID.
+	XStorjRequestID = "X-Storj-Request-Id"
+)
+
+// AddRequestID uses XStorjRequestId field to set unique request Ids
+// in the response headers for each request of auth and linksharing service, if it dosen't alreasy exist.
+func AddRequestID(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
```
Comment (halkyon) [inline pkg/middleware/requestId.go:26]:
```suggestion
requestID := r.Header.Get(XStorjRequestID)
if requestID != "" {
  requestID = fmt.Sprintf("%x", monkit.NewId())
}

w.Header().Set(XStorjRequestID, requestID)
ctx := r.Context().WithValue(ctx, RequestIDKey, requestID)
h.ServeHTTP(w, r.WithContext(ctx))
```

**Review by halkyon — APPROVED** [review]

File: pkg/middleware/requestId.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	if reqID, ok := ctx.Value(RequestIDKey{}).(string); ok {
+		return reqID
+	}
+	return ""
+}
+
+// AddReqIDHeader adds the request ID from the context to the request header.
+func AddReqIDHeader(req *http.Request) {
```
Comment (amwolff) [inline pkg/middleware/requestId.go:54]: nit: I think we can use full naming here without abbreviations, i.e., `AddRequestIDHeader` (or `AddRequestIDToHeaders`).

**Review by amwolff — COMMENTED** [review]
I have a few stylistic suggestions. Otherwise, it looks good to me!

**Review by amwolff — APPROVED** [review]
Good work, Aishwarya! 🥇

**Review by amwolff — APPROVED** [review]

**Review by pwilloughby — APPROVED** [review]
